{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import requests\n",
    "import pathlib\n",
    "from requests.auth import HTTPBasicAuth\n",
    "\n",
    "api_key = os.environ.get(\"NCBI_API_KEY\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input\n",
    "taxon_file = \"../data/imported/taxids.list\"\n",
    "\n",
    "# Output\n",
    "output_dir = pathlib.Path(\"../data/generated/download_genomes\")\n",
    "pathlib.Path.mkdir(output_dir, exist_ok=True)\n",
    "genomes_metadata_file = output_dir / \"genomes_metadata.csv\"\n",
    "genome_accessions_file = output_dir / \"genome_accessions.list\"\n",
    "genomes_zip_file = output_dir / \"genomes.zip\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genome_dataset_report(taxon_id, api_key, params=None):\n",
    "    \"\"\"\n",
    "    Get the genome dataset report for a given taxon id using NCBI API.\n",
    "    Default parameters are:\n",
    "    - reference_only=true\n",
    "    - assembly_source=refseq\n",
    "    - exclude_atypical=true # => don't include genomes that have assembly issues or are otherwise atypical\n",
    "    - assembly_version=current\n",
    "    \"\"\"\n",
    "    if params is None:\n",
    "        params = {\n",
    "            \"filters.reference_only\": \"true\",\n",
    "            \"filters.assembly_source\": \"refseq\",\n",
    "            \"filters.exclude_atypical\": \"true\",\n",
    "            \"filters.assembly_version\": \"current\",\n",
    "        }\n",
    "    auth = HTTPBasicAuth(\"api-key\", api_key)\n",
    "    url = f\"https://api.ncbi.nlm.nih.gov/datasets/v2alpha/genome/taxon/{taxon_id}/dataset_report\"\n",
    "    response = requests.get(url=url, auth=auth, params=params)\n",
    "    if \"reports\" not in response.json():\n",
    "        print(f\"No reports found for taxon id {taxon_id}\")\n",
    "        return\n",
    "    return pd.DataFrame(\n",
    "        _parse_genome_dataset_report_response(response.json(), taxon_id)\n",
    "    )\n",
    "\n",
    "\n",
    "def _parse_genome_dataset_report_response(response, query_taxid):\n",
    "    for report in response[\"reports\"]:\n",
    "        yield {\n",
    "            \"query_taxid\": query_taxid,\n",
    "            \"genome_accession\": report[\"current_accession\"],\n",
    "            \"genome_name\": report[\"organism\"][\"organism_name\"],\n",
    "            \"genome_taxid\": report[\"organism\"][\"tax_id\"],\n",
    "            \"assembly_level\": report[\"assembly_info\"][\"assembly_level\"],\n",
    "            \"assembly_status\": report[\"assembly_info\"][\"assembly_status\"],\n",
    "            \"assembly_name\": report[\"assembly_info\"][\"assembly_name\"],\n",
    "            \"refseq_category\": report[\"assembly_info\"][\"refseq_category\"]\n",
    "            if \"refseq_category\" in report[\"assembly_info\"]\n",
    "            else pd.NA,\n",
    "            \"number_of_contigs\": report[\"assembly_stats\"][\"number_of_contigs\"],\n",
    "            \"contig_n50\": report[\"assembly_stats\"][\"contig_n50\"],\n",
    "            \"contig_l50\": report[\"assembly_stats\"][\"contig_l50\"],\n",
    "            \"genome_coverage\": float(\n",
    "                str(report[\"assembly_stats\"][\"genome_coverage\"]).replace(\"x\", \"\")\n",
    "            )\n",
    "            if \"genome_coverage\" in report[\"assembly_stats\"]\n",
    "            else pd.NA,\n",
    "            \"total_sequence_length\": report[\"assembly_stats\"][\"total_sequence_length\"],\n",
    "            \"total_ungapped_length\": report[\"assembly_stats\"][\"total_ungapped_length\"],\n",
    "        }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>562</td>\n",
       "      <td>GCF_000005845.2</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>511145</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM584v2</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>1</td>\n",
       "      <td>4641652</td>\n",
       "      <td>1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>4641652</td>\n",
       "      <td>4641652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>562</td>\n",
       "      <td>GCF_000008865.2</td>\n",
       "      <td>Escherichia coli O157:H7 str. Sakai</td>\n",
       "      <td>386585</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM886v2</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>3</td>\n",
       "      <td>5498578</td>\n",
       "      <td>1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>5594605</td>\n",
       "      <td>5594605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  query_taxid genome_accession                                genome_name  \\\n",
       "0         562  GCF_000005845.2  Escherichia coli str. K-12 substr. MG1655   \n",
       "1         562  GCF_000008865.2        Escherichia coli O157:H7 str. Sakai   \n",
       "\n",
       "   genome_taxid   assembly_level assembly_status assembly_name  \\\n",
       "0        511145  Complete Genome         current      ASM584v2   \n",
       "1        386585  Complete Genome         current      ASM886v2   \n",
       "\n",
       "    refseq_category  number_of_contigs  contig_n50  contig_l50  \\\n",
       "0  reference genome                  1     4641652           1   \n",
       "1  reference genome                  3     5498578           1   \n",
       "\n",
       "  genome_coverage total_sequence_length total_ungapped_length  \n",
       "0            <NA>               4641652               4641652  \n",
       "1            <NA>               5594605               5594605  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_params = {\n",
    "    \"filters.reference_only\": \"true\",\n",
    "    \"filters.assembly_source\": \"refseq\",\n",
    "    \"filters.exclude_atypical\": \"true\",\n",
    "    \"filters.assembly_version\": \"current\",\n",
    "}\n",
    "\n",
    "genome_dataset_report = get_genome_dataset_report(\"562\", api_key, params=refseq_params)\n",
    "genome_dataset_report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "162 taxon ids read from ../data/imported/taxids.list\n"
     ]
    }
   ],
   "source": [
    "def read_taxon_list(taxon_file):\n",
    "    \"\"\"\n",
    "    Read a list of taxon ids from a file.\n",
    "    \"\"\"\n",
    "    taxon_list = []\n",
    "    with open(taxon_file) as f:\n",
    "        taxon_list.extend(line.strip() for line in f if not line.startswith(\"#\"))\n",
    "    return sorted(set(taxon_list))\n",
    "\n",
    "\n",
    "taxon_list = read_taxon_list(taxon_file)\n",
    "print(f\"{len(taxon_list)} taxon ids read from {taxon_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No reports found for taxon id 1235797\n",
      "No reports found for taxon id 142586\n",
      "No reports found for taxon id 1454229\n",
      "No reports found for taxon id 1686296\n",
      "No reports found for taxon id 1849041\n",
      "No reports found for taxon id 1935934\n",
      "No reports found for taxon id 1955243\n",
      "No reports found for taxon id 1965293\n",
      "No reports found for taxon id 2039241\n",
      "No reports found for taxon id 2040332\n",
      "No reports found for taxon id 2049021\n",
      "No reports found for taxon id 39491\n",
      "No reports found for taxon id 39492\n",
      "No reports found for taxon id 41978\n",
      "No reports found for taxon id 885\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>626932</td>\n",
       "      <td>GCF_014163495.1</td>\n",
       "      <td>Alistipes indistinctus</td>\n",
       "      <td>626932</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM1416349v1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>2</td>\n",
       "      <td>3012737</td>\n",
       "      <td>1</td>\n",
       "      <td>196.0</td>\n",
       "      <td>3095585</td>\n",
       "      <td>3095585</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    query_taxid genome_accession             genome_name  genome_taxid  \\\n",
       "129      626932  GCF_014163495.1  Alistipes indistinctus        626932   \n",
       "\n",
       "      assembly_level assembly_status assembly_name        refseq_category  \\\n",
       "129  Complete Genome         current  ASM1416349v1  representative genome   \n",
       "\n",
       "     number_of_contigs  contig_n50  contig_l50 genome_coverage  \\\n",
       "129                  2     3012737           1           196.0   \n",
       "\n",
       "    total_sequence_length total_ungapped_length  \n",
       "129               3095585               3095585  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_genomes_dataset_reports = [\n",
    "    get_genome_dataset_report(taxon_id, api_key, params=refseq_params)\n",
    "    for taxon_id in taxon_list\n",
    "]\n",
    "refseq_genomes_df = pd.concat(refseq_genomes_dataset_reports).reset_index(drop=True)\n",
    "refseq_genomes_df.sample()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "representative genome    147\n",
       "Name: refseq_category, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_genomes_df[\"refseq_category\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Complete Genome    70\n",
       "Scaffold           43\n",
       "Contig             25\n",
       "Chromosome          9\n",
       "Name: assembly_level, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_genomes_df[\"assembly_level\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15 taxon ids failed to retrieve genome dataset report\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['1935934',\n",
       " '1965293',\n",
       " '1454229',\n",
       " '39491',\n",
       " '41978',\n",
       " '885',\n",
       " '2039241',\n",
       " '39492',\n",
       " '2040332',\n",
       " '1955243',\n",
       " '1235797',\n",
       " '142586',\n",
       " '1686296',\n",
       " '2049021',\n",
       " '1849041']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "failed_taxon_ids = list(\n",
    "    set(taxon_list).difference(set(refseq_genomes_df[\"query_taxid\"]))\n",
    ")\n",
    "print(f\"{len(failed_taxon_ids)} taxon ids failed to retrieve genome dataset report\")\n",
    "failed_taxon_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No reports found for taxon id 1849041\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>893</th>\n",
       "      <td>39492</td>\n",
       "      <td>GCA_000210635.1</td>\n",
       "      <td>[Eubacterium] siraeum V10Sc8a</td>\n",
       "      <td>717961</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM21063v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>107</td>\n",
       "      <td>46087</td>\n",
       "      <td>18</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>2836123</td>\n",
       "      <td>2779231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1241</th>\n",
       "      <td>142586</td>\n",
       "      <td>GCA_017633905.1</td>\n",
       "      <td>Eubacterium sp.</td>\n",
       "      <td>142586</td>\n",
       "      <td>Contig</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM1763390v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>166</td>\n",
       "      <td>12613</td>\n",
       "      <td>43</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1754401</td>\n",
       "      <td>1754401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>41978</td>\n",
       "      <td>GCA_017404665.1</td>\n",
       "      <td>Ruminococcus sp.</td>\n",
       "      <td>41978</td>\n",
       "      <td>Contig</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM1740466v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>79</td>\n",
       "      <td>49163</td>\n",
       "      <td>14</td>\n",
       "      <td>30.0</td>\n",
       "      <td>2465842</td>\n",
       "      <td>2465842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>39491</td>\n",
       "      <td>GCA_022738285.1</td>\n",
       "      <td>[Eubacterium] rectale</td>\n",
       "      <td>39491</td>\n",
       "      <td>Contig</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM2273828v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>535</td>\n",
       "      <td>7235</td>\n",
       "      <td>109</td>\n",
       "      <td>434.4</td>\n",
       "      <td>2593202</td>\n",
       "      <td>2593202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1303</th>\n",
       "      <td>142586</td>\n",
       "      <td>GCA_021631215.1</td>\n",
       "      <td>Eubacterium sp.</td>\n",
       "      <td>142586</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM2163121v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>749</td>\n",
       "      <td>2885</td>\n",
       "      <td>201</td>\n",
       "      <td>8.4</td>\n",
       "      <td>1857763</td>\n",
       "      <td>1857493</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     query_taxid genome_accession                    genome_name  \\\n",
       "893        39492  GCA_000210635.1  [Eubacterium] siraeum V10Sc8a   \n",
       "1241      142586  GCA_017633905.1                Eubacterium sp.   \n",
       "405        41978  GCA_017404665.1               Ruminococcus sp.   \n",
       "195        39491  GCA_022738285.1          [Eubacterium] rectale   \n",
       "1303      142586  GCA_021631215.1                Eubacterium sp.   \n",
       "\n",
       "      genome_taxid assembly_level assembly_status assembly_name  \\\n",
       "893         717961     Chromosome         current    ASM21063v1   \n",
       "1241        142586         Contig         current  ASM1763390v1   \n",
       "405          41978         Contig         current  ASM1740466v1   \n",
       "195          39491         Contig         current  ASM2273828v1   \n",
       "1303        142586       Scaffold         current  ASM2163121v1   \n",
       "\n",
       "     refseq_category  number_of_contigs  contig_n50  contig_l50  \\\n",
       "893             <NA>                107       46087          18   \n",
       "1241            <NA>                166       12613          43   \n",
       "405             <NA>                 79       49163          14   \n",
       "195             <NA>                535        7235         109   \n",
       "1303            <NA>                749        2885         201   \n",
       "\n",
       "     genome_coverage total_sequence_length total_ungapped_length  \n",
       "893             <NA>               2836123               2779231  \n",
       "1241            30.0               1754401               1754401  \n",
       "405             30.0               2465842               2465842  \n",
       "195            434.4               2593202               2593202  \n",
       "1303             8.4               1857763               1857493  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank_params = {\n",
    "    \"filters.assembly_source\": \"genbank\",\n",
    "    \"filters.exclude_atypical\": \"true\",\n",
    "    \"filters.assembly_version\": \"current\",\n",
    "    \"table_fields\": \"assminfo-accession,assminfo-name\",\n",
    "    \"page_size\": 500,\n",
    "}\n",
    "\n",
    "genbank_genome_dataset_reports = [\n",
    "    get_genome_dataset_report(failed_taxon_id, api_key, params=genbank_params)\n",
    "    for failed_taxon_id in failed_taxon_ids\n",
    "]\n",
    "genbank_genomes_df = pd.concat(genbank_genome_dataset_reports).reset_index(drop=True)\n",
    "genbank_genomes_df.sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Contig             973\n",
       "Scaffold           481\n",
       "Chromosome           8\n",
       "Complete Genome      3\n",
       "Name: assembly_level, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank_genomes_df[\"assembly_level\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "      <th>gaps</th>\n",
       "      <th>bases_per_contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1454229</td>\n",
       "      <td>GCA_014385295.1</td>\n",
       "      <td>Bifidobacterium faecale</td>\n",
       "      <td>1454229</td>\n",
       "      <td>Contig</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM1438529v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>24</td>\n",
       "      <td>1422868</td>\n",
       "      <td>1</td>\n",
       "      <td>62.0</td>\n",
       "      <td>2257479</td>\n",
       "      <td>2257479</td>\n",
       "      <td>0</td>\n",
       "      <td>94061.625</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   query_taxid genome_accession              genome_name  genome_taxid  \\\n",
       "6      1454229  GCA_014385295.1  Bifidobacterium faecale       1454229   \n",
       "\n",
       "  assembly_level assembly_status assembly_name refseq_category  \\\n",
       "6         Contig         current  ASM1438529v1            <NA>   \n",
       "\n",
       "   number_of_contigs  contig_n50  contig_l50  genome_coverage  \\\n",
       "6                 24     1422868           1             62.0   \n",
       "\n",
       "   total_sequence_length  total_ungapped_length  gaps  bases_per_contig  \n",
       "6                2257479                2257479     0         94061.625  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def select_best_representative(df, consider_coverage=None):\n",
    "    \"\"\"\n",
    "    Select the best representative genome from a genome dataset report.\n",
    "    \"\"\"\n",
    "    if consider_coverage is None:\n",
    "        consider_coverage = True\n",
    "    taxa_dfs = []\n",
    "    # convert columns to correct data types\n",
    "    df = df.astype(\n",
    "        {\n",
    "            \"query_taxid\": \"int\",\n",
    "            \"genome_taxid\": \"int\",\n",
    "            \"number_of_contigs\": \"int\",\n",
    "            \"total_sequence_length\": \"int\",\n",
    "            \"total_ungapped_length\": \"int\",\n",
    "            \"contig_n50\": \"int\",\n",
    "            \"contig_l50\": \"int\",\n",
    "            # \"genome_coverage\": \"float\",\n",
    "        }\n",
    "    )\n",
    "    df[\"gaps\"] = df[\"total_sequence_length\"] - df[\"total_ungapped_length\"]\n",
    "    df[\"bases_per_contig\"] = df[\"total_ungapped_length\"] / df[\"number_of_contigs\"]\n",
    "\n",
    "    if consider_coverage:\n",
    "        # if genome coverage is NA remove it from the dataframe\n",
    "        df = df.dropna(subset=[\"genome_coverage\"]).copy()\n",
    "\n",
    "    # Select the representative genome with the highest genome coverage, least gaps and least contig_l50 and highest contig_n50.\n",
    "    for _, grouped_taxa_df in df.groupby(\"query_taxid\"):\n",
    "        rows = grouped_taxa_df.shape[0]\n",
    "        if rows == 1:\n",
    "            taxa_dfs.append(grouped_taxa_df.iloc[0])\n",
    "            continue\n",
    "\n",
    "        # if a taxon has any 'Complete Genome's then shortlist those genomes to be considered for a representative genome.\n",
    "        if any(grouped_taxa_df[\"assembly_level\"] == \"Complete Genome\"):\n",
    "            grouped_taxa_df = grouped_taxa_df.query(\n",
    "                \"assembly_level == 'Complete Genome'\"\n",
    "            ).copy()\n",
    "\n",
    "        # Select the representative genome with the highest genome coverage, least gaps and least contig_l50 and highest contig_n50.\n",
    "        grouped_taxa_df = grouped_taxa_df.sort_values(\n",
    "            [\"bases_per_contig\", \"contig_n50\", \"contig_l50\", \"genome_coverage\", \"gaps\"],\n",
    "            ascending=[False, False, True, False, True],\n",
    "        )\n",
    "\n",
    "        # append the best representative genome to the taxa_dfs list\n",
    "        taxa_dfs.append(grouped_taxa_df.iloc[0])\n",
    "\n",
    "    return pd.DataFrame(taxa_dfs)\n",
    "\n",
    "\n",
    "select_best_representative(genbank_genomes_df).reset_index(drop=True).sample()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "      <th>gaps</th>\n",
       "      <th>bases_per_contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>1454229</td>\n",
       "      <td>GCA_014385295.1</td>\n",
       "      <td>Bifidobacterium faecale</td>\n",
       "      <td>1454229</td>\n",
       "      <td>Contig</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM1438529v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>24</td>\n",
       "      <td>1422868</td>\n",
       "      <td>1</td>\n",
       "      <td>62.0</td>\n",
       "      <td>2257479</td>\n",
       "      <td>2257479</td>\n",
       "      <td>0</td>\n",
       "      <td>9.406162e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1735</td>\n",
       "      <td>GCF_003464225.1</td>\n",
       "      <td>Holdemanella biformis</td>\n",
       "      <td>1735</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM346422v1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>87</td>\n",
       "      <td>63083</td>\n",
       "      <td>12</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2477128</td>\n",
       "      <td>2476939</td>\n",
       "      <td>189</td>\n",
       "      <td>2.847056e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>47678</td>\n",
       "      <td>GCF_018292205.1</td>\n",
       "      <td>Bacteroides caccae CL03T12C61</td>\n",
       "      <td>997873</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM1829220v1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>4</td>\n",
       "      <td>5485670</td>\n",
       "      <td>1</td>\n",
       "      <td>164.0</td>\n",
       "      <td>5499725</td>\n",
       "      <td>5499725</td>\n",
       "      <td>0</td>\n",
       "      <td>1.374931e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>2039241</td>\n",
       "      <td>GCA_020024265.1</td>\n",
       "      <td>Anaerotignum sp.</td>\n",
       "      <td>2039241</td>\n",
       "      <td>Contig</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM2002426v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>50</td>\n",
       "      <td>62090</td>\n",
       "      <td>10</td>\n",
       "      <td>66.4</td>\n",
       "      <td>1877045</td>\n",
       "      <td>1877045</td>\n",
       "      <td>0</td>\n",
       "      <td>3.754090e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>39486</td>\n",
       "      <td>GCF_000169235.1</td>\n",
       "      <td>Dorea formicigenerans ATCC 27755</td>\n",
       "      <td>411461</td>\n",
       "      <td>Contig</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM16923v1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>16</td>\n",
       "      <td>318591</td>\n",
       "      <td>3</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>3186031</td>\n",
       "      <td>3186031</td>\n",
       "      <td>0</td>\n",
       "      <td>1.991269e+05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     query_taxid genome_accession                       genome_name  \\\n",
       "153      1454229  GCA_014385295.1           Bifidobacterium faecale   \n",
       "35          1735  GCF_003464225.1             Holdemanella biformis   \n",
       "65         47678  GCF_018292205.1     Bacteroides caccae CL03T12C61   \n",
       "158      2039241  GCA_020024265.1                  Anaerotignum sp.   \n",
       "54         39486  GCF_000169235.1  Dorea formicigenerans ATCC 27755   \n",
       "\n",
       "     genome_taxid   assembly_level assembly_status assembly_name  \\\n",
       "153       1454229           Contig         current  ASM1438529v1   \n",
       "35           1735         Scaffold         current   ASM346422v1   \n",
       "65         997873  Complete Genome         current  ASM1829220v1   \n",
       "158       2039241           Contig         current  ASM2002426v1   \n",
       "54         411461           Contig         current    ASM16923v1   \n",
       "\n",
       "           refseq_category  number_of_contigs  contig_n50  contig_l50  \\\n",
       "153                   <NA>                 24     1422868           1   \n",
       "35   representative genome                 87       63083          12   \n",
       "65   representative genome                  4     5485670           1   \n",
       "158                   <NA>                 50       62090          10   \n",
       "54   representative genome                 16      318591           3   \n",
       "\n",
       "    genome_coverage  total_sequence_length  total_ungapped_length  gaps  \\\n",
       "153            62.0                2257479                2257479     0   \n",
       "35            100.0                2477128                2476939   189   \n",
       "65            164.0                5499725                5499725     0   \n",
       "158            66.4                1877045                1877045     0   \n",
       "54             <NA>                3186031                3186031     0   \n",
       "\n",
       "     bases_per_contig  \n",
       "153      9.406162e+04  \n",
       "35       2.847056e+04  \n",
       "65       1.374931e+06  \n",
       "158      3.754090e+04  \n",
       "54       1.991269e+05  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomes_df = pd.concat(\n",
    "    [\n",
    "        select_best_representative(refseq_genomes_df, consider_coverage=False),\n",
    "        select_best_representative(genbank_genomes_df),\n",
    "    ]\n",
    ").reset_index(drop=True)\n",
    "genomes_df.to_csv(genomes_metadata_file, index=False)\n",
    "genomes_df.sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GCF_016889925.1    1\n",
       "GCF_009721605.1    1\n",
       "GCF_020341675.1    1\n",
       "GCF_025149125.1    1\n",
       "GCF_008571395.1    1\n",
       "                  ..\n",
       "GCF_018140635.1    1\n",
       "GCF_003475105.1    1\n",
       "GCF_902375065.1    1\n",
       "GCF_900637515.1    1\n",
       "GCA_028727085.1    1\n",
       "Name: genome_accession, Length: 161, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomes_df[\"genome_accession\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Check to make sure there are no duplicate genome accessions\n",
    "assert all(\n",
    "    genomes_df[\"genome_accession\"].value_counts() == 1\n",
    "), \"Duplicate genome accessions found!\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the genome accessions to a file\n",
    "genomes_df[\"genome_accession\"].to_csv(genome_accessions_file, index=False, header=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Manually added `GCF_013304735.1` and `GCF_013305105.1` to accessions list as substitutes for the two Blautia lutia genomes (Blautia sp. (95% luti) and Blautia sp. (97% luti))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No genomes could be found for NCBI TaxID `1849041`, but we have it assembled here: `s3://genomics-workflow-core/Results/HybridAssembly/MITI-MCB/SH0001499-00107/UNICYCLER/assembly.fasta`\n",
    "So I'll add this to the list of genomes to be processed."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download the genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$api_key\" \"$genome_accessions_file\" \"$genomes_zip_file\"\n",
    "\n",
    "datasets download genome accession \\\n",
    "    --api-key $1 \\\n",
    "    --inputfile $2 \\\n",
    "    --no-progressbar \\\n",
    "    --filename $3\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "trees",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
