{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import requests\n",
    "import pathlib\n",
    "from requests.auth import HTTPBasicAuth\n",
    "\n",
    "api_key = os.environ.get(\"NCBI_API_KEY\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input\n",
    "taxon_file = \"../data/imported/taxids.list\"\n",
    "\n",
    "# Output\n",
    "output_dir = pathlib.Path(\"../data/generated/download_genomes\")\n",
    "pathlib.Path.mkdir(output_dir, exist_ok=True)\n",
    "genomes_metadata_file = output_dir / \"genomes_metadata.csv\"\n",
    "genome_accessions_file = output_dir / \"genome_accessions.list\"\n",
    "genomes_zip_file = output_dir / \"genomes.zip\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genome_dataset_report(taxon_id, api_key, params=None):\n",
    "    \"\"\"\n",
    "    Get the genome dataset report for a given taxon id using NCBI API.\n",
    "    Default parameters are:\n",
    "    - reference_only=true\n",
    "    - assembly_source=refseq\n",
    "    - exclude_atypical=true # => don't include genomes that have assembly issues or are otherwise atypical\n",
    "    - assembly_version=current\n",
    "    \"\"\"\n",
    "    if params is None:\n",
    "        params = {\n",
    "            \"filters.reference_only\": \"true\",\n",
    "            \"filters.assembly_source\": \"refseq\",\n",
    "            \"filters.exclude_atypical\": \"true\",\n",
    "            \"filters.assembly_version\": \"current\",\n",
    "        }\n",
    "    auth = HTTPBasicAuth(\"api-key\", api_key)\n",
    "    url = f\"https://api.ncbi.nlm.nih.gov/datasets/v2alpha/genome/taxon/{taxon_id}/dataset_report\"\n",
    "    response = requests.get(url=url, auth=auth, params=params)\n",
    "    if \"reports\" not in response.json():\n",
    "        print(f\"No reports found for taxon id {taxon_id}\")\n",
    "        return\n",
    "    return pd.DataFrame(\n",
    "        _parse_genome_dataset_report_response(response.json(), taxon_id)\n",
    "    )\n",
    "\n",
    "\n",
    "def _parse_genome_dataset_report_response(response, query_taxid):\n",
    "    for report in response[\"reports\"]:\n",
    "        yield {\n",
    "            \"query_taxid\": query_taxid,\n",
    "            \"genome_accession\": report[\"current_accession\"],\n",
    "            \"genome_name\": report[\"organism\"][\"organism_name\"],\n",
    "            \"genome_taxid\": report[\"organism\"][\"tax_id\"],\n",
    "            \"assembly_level\": report[\"assembly_info\"][\"assembly_level\"],\n",
    "            \"assembly_status\": report[\"assembly_info\"][\"assembly_status\"],\n",
    "            \"assembly_name\": report[\"assembly_info\"][\"assembly_name\"],\n",
    "            \"refseq_category\": report[\"assembly_info\"][\"refseq_category\"]\n",
    "            if \"refseq_category\" in report[\"assembly_info\"]\n",
    "            else pd.NA,\n",
    "            \"number_of_contigs\": report[\"assembly_stats\"][\"number_of_contigs\"],\n",
    "            \"contig_n50\": report[\"assembly_stats\"][\"contig_n50\"],\n",
    "            \"contig_l50\": report[\"assembly_stats\"][\"contig_l50\"],\n",
    "            \"genome_coverage\": float(\n",
    "                str(report[\"assembly_stats\"][\"genome_coverage\"]).replace(\"x\", \"\")\n",
    "            )\n",
    "            if \"genome_coverage\" in report[\"assembly_stats\"]\n",
    "            else pd.NA,\n",
    "            \"total_sequence_length\": report[\"assembly_stats\"][\"total_sequence_length\"],\n",
    "            \"total_ungapped_length\": report[\"assembly_stats\"][\"total_ungapped_length\"],\n",
    "        }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>562</td>\n",
       "      <td>GCF_000005845.2</td>\n",
       "      <td>Escherichia coli str. K-12 substr. MG1655</td>\n",
       "      <td>511145</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM584v2</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>1</td>\n",
       "      <td>4641652</td>\n",
       "      <td>1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>4641652</td>\n",
       "      <td>4641652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>562</td>\n",
       "      <td>GCF_000008865.2</td>\n",
       "      <td>Escherichia coli O157:H7 str. Sakai</td>\n",
       "      <td>386585</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM886v2</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>3</td>\n",
       "      <td>5498578</td>\n",
       "      <td>1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>5594605</td>\n",
       "      <td>5594605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  query_taxid genome_accession                                genome_name  \\\n",
       "0         562  GCF_000005845.2  Escherichia coli str. K-12 substr. MG1655   \n",
       "1         562  GCF_000008865.2        Escherichia coli O157:H7 str. Sakai   \n",
       "\n",
       "   genome_taxid   assembly_level assembly_status assembly_name  \\\n",
       "0        511145  Complete Genome         current      ASM584v2   \n",
       "1        386585  Complete Genome         current      ASM886v2   \n",
       "\n",
       "    refseq_category  number_of_contigs  contig_n50  contig_l50  \\\n",
       "0  reference genome                  1     4641652           1   \n",
       "1  reference genome                  3     5498578           1   \n",
       "\n",
       "  genome_coverage total_sequence_length total_ungapped_length  \n",
       "0            <NA>               4641652               4641652  \n",
       "1            <NA>               5594605               5594605  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_params = {\n",
    "    \"filters.reference_only\": \"true\",\n",
    "    \"filters.assembly_source\": \"refseq\",\n",
    "    \"filters.exclude_atypical\": \"true\",\n",
    "    \"filters.assembly_version\": \"current\",\n",
    "}\n",
    "\n",
    "genome_dataset_report = get_genome_dataset_report(\"562\", api_key, params=refseq_params)\n",
    "genome_dataset_report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "161 taxon ids read from ../data/imported/taxids.list\n"
     ]
    }
   ],
   "source": [
    "def read_taxon_list(taxon_file):\n",
    "    \"\"\"\n",
    "    Read a list of taxon ids from a file.\n",
    "    \"\"\"\n",
    "    taxon_list = []\n",
    "    with open(taxon_file) as f:\n",
    "        taxon_list.extend(line.strip() for line in f if not line.startswith(\"#\"))\n",
    "    return sorted(set(taxon_list))\n",
    "\n",
    "\n",
    "taxon_list = read_taxon_list(taxon_file)\n",
    "print(f\"{len(taxon_list)} taxon ids read from {taxon_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No reports found for taxon id 1235797\n",
      "No reports found for taxon id 142586\n",
      "No reports found for taxon id 1454229\n",
      "No reports found for taxon id 1686296\n",
      "No reports found for taxon id 1849041\n",
      "No reports found for taxon id 1935934\n",
      "No reports found for taxon id 1955243\n",
      "No reports found for taxon id 1965293\n",
      "No reports found for taxon id 2039241\n",
      "No reports found for taxon id 2040332\n",
      "No reports found for taxon id 2049021\n",
      "No reports found for taxon id 39491\n",
      "No reports found for taxon id 39492\n",
      "No reports found for taxon id 41978\n",
      "No reports found for taxon id 885\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>61171</td>\n",
       "      <td>GCF_003459085.1</td>\n",
       "      <td>Holdemania filiformis</td>\n",
       "      <td>61171</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM345908v1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>58</td>\n",
       "      <td>141100</td>\n",
       "      <td>9</td>\n",
       "      <td>100.0</td>\n",
       "      <td>3744856</td>\n",
       "      <td>3744494</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    query_taxid genome_accession            genome_name  genome_taxid  \\\n",
       "128       61171  GCF_003459085.1  Holdemania filiformis         61171   \n",
       "\n",
       "    assembly_level assembly_status assembly_name        refseq_category  \\\n",
       "128       Scaffold         current   ASM345908v1  representative genome   \n",
       "\n",
       "     number_of_contigs  contig_n50  contig_l50 genome_coverage  \\\n",
       "128                 58      141100           9           100.0   \n",
       "\n",
       "    total_sequence_length total_ungapped_length  \n",
       "128               3744856               3744494  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_genomes_dataset_reports = [\n",
    "    get_genome_dataset_report(taxon_id, api_key, params=refseq_params)\n",
    "    for taxon_id in taxon_list\n",
    "]\n",
    "refseq_genomes_df = pd.concat(refseq_genomes_dataset_reports).reset_index(drop=True)\n",
    "refseq_genomes_df.sample()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "representative genome    146\n",
       "Name: refseq_category, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_genomes_df[\"refseq_category\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Complete Genome    70\n",
       "Scaffold           42\n",
       "Contig             25\n",
       "Chromosome          9\n",
       "Name: assembly_level, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_genomes_df[\"assembly_level\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15 taxon ids failed to retrieve genome dataset report\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['2040332',\n",
       " '1955243',\n",
       " '1235797',\n",
       " '1686296',\n",
       " '1849041',\n",
       " '142586',\n",
       " '885',\n",
       " '1935934',\n",
       " '39491',\n",
       " '1965293',\n",
       " '39492',\n",
       " '2039241',\n",
       " '41978',\n",
       " '2049021',\n",
       " '1454229']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "failed_taxon_ids = list(\n",
    "    set(taxon_list).difference(set(refseq_genomes_df[\"query_taxid\"]))\n",
    ")\n",
    "print(f\"{len(failed_taxon_ids)} taxon ids failed to retrieve genome dataset report\")\n",
    "failed_taxon_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No reports found for taxon id 1849041\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>849</th>\n",
       "      <td>39492</td>\n",
       "      <td>GCA_000209915.1</td>\n",
       "      <td>[Eubacterium] siraeum 70/3</td>\n",
       "      <td>657319</td>\n",
       "      <td>Chromosome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM20991v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>139</td>\n",
       "      <td>38521</td>\n",
       "      <td>22</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>2943413</td>\n",
       "      <td>2864137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>643</th>\n",
       "      <td>39491</td>\n",
       "      <td>GCA_003469095.1</td>\n",
       "      <td>[Eubacterium] rectale</td>\n",
       "      <td>39491</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM346909v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>135</td>\n",
       "      <td>122438</td>\n",
       "      <td>8</td>\n",
       "      <td>100.0</td>\n",
       "      <td>3243530</td>\n",
       "      <td>3242497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>833</th>\n",
       "      <td>39491</td>\n",
       "      <td>GCA_934871245.1</td>\n",
       "      <td>[Eubacterium] rectale</td>\n",
       "      <td>39491</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>current</td>\n",
       "      <td>MTG238_bin.4.fa</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>65</td>\n",
       "      <td>88437</td>\n",
       "      <td>9</td>\n",
       "      <td>20.0</td>\n",
       "      <td>2738878</td>\n",
       "      <td>2738178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>2040332</td>\n",
       "      <td>GCA_025541975.1</td>\n",
       "      <td>Dorea sp.</td>\n",
       "      <td>2040332</td>\n",
       "      <td>Contig</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM2554197v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>47</td>\n",
       "      <td>128628</td>\n",
       "      <td>5</td>\n",
       "      <td>2.6</td>\n",
       "      <td>2259685</td>\n",
       "      <td>2259685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>142586</td>\n",
       "      <td>GCA_022800015.1</td>\n",
       "      <td>Eubacterium sp.</td>\n",
       "      <td>142586</td>\n",
       "      <td>Contig</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM2280001v1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>149</td>\n",
       "      <td>30263</td>\n",
       "      <td>28</td>\n",
       "      <td>73.0</td>\n",
       "      <td>2954030</td>\n",
       "      <td>2954030</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    query_taxid genome_accession                 genome_name  genome_taxid  \\\n",
       "849       39492  GCA_000209915.1  [Eubacterium] siraeum 70/3        657319   \n",
       "643       39491  GCA_003469095.1       [Eubacterium] rectale         39491   \n",
       "833       39491  GCA_934871245.1       [Eubacterium] rectale         39491   \n",
       "43      2040332  GCA_025541975.1                   Dorea sp.       2040332   \n",
       "436      142586  GCA_022800015.1             Eubacterium sp.        142586   \n",
       "\n",
       "    assembly_level assembly_status    assembly_name refseq_category  \\\n",
       "849     Chromosome         current       ASM20991v1            <NA>   \n",
       "643       Scaffold         current      ASM346909v1            <NA>   \n",
       "833       Scaffold         current  MTG238_bin.4.fa            <NA>   \n",
       "43          Contig         current     ASM2554197v1            <NA>   \n",
       "436         Contig         current     ASM2280001v1            <NA>   \n",
       "\n",
       "     number_of_contigs  contig_n50  contig_l50 genome_coverage  \\\n",
       "849                139       38521          22            <NA>   \n",
       "643                135      122438           8           100.0   \n",
       "833                 65       88437           9            20.0   \n",
       "43                  47      128628           5             2.6   \n",
       "436                149       30263          28            73.0   \n",
       "\n",
       "    total_sequence_length total_ungapped_length  \n",
       "849               2943413               2864137  \n",
       "643               3243530               3242497  \n",
       "833               2738878               2738178  \n",
       "43                2259685               2259685  \n",
       "436               2954030               2954030  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank_params = {\n",
    "    \"filters.assembly_source\": \"genbank\",\n",
    "    \"filters.exclude_atypical\": \"true\",\n",
    "    \"filters.assembly_version\": \"current\",\n",
    "    \"table_fields\": \"assminfo-accession,assminfo-name\",\n",
    "    \"page_size\": 500,\n",
    "}\n",
    "\n",
    "genbank_genome_dataset_reports = [\n",
    "    get_genome_dataset_report(failed_taxon_id, api_key, params=genbank_params)\n",
    "    for failed_taxon_id in failed_taxon_ids\n",
    "]\n",
    "genbank_genomes_df = pd.concat(genbank_genome_dataset_reports).reset_index(drop=True)\n",
    "genbank_genomes_df.sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Contig             973\n",
       "Scaffold           481\n",
       "Chromosome           8\n",
       "Complete Genome      3\n",
       "Name: assembly_level, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank_genomes_df[\"assembly_level\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "      <th>gaps</th>\n",
       "      <th>bases_per_contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1235797</td>\n",
       "      <td>GCA_000403435.2</td>\n",
       "      <td>Oscillibacter sp. 1-3</td>\n",
       "      <td>1235797</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>current</td>\n",
       "      <td>Osci_bact_1-3_V1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>43</td>\n",
       "      <td>271557</td>\n",
       "      <td>6</td>\n",
       "      <td>88.0</td>\n",
       "      <td>4467686</td>\n",
       "      <td>4430205</td>\n",
       "      <td>37481</td>\n",
       "      <td>103028.023256</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   query_taxid genome_accession            genome_name  genome_taxid  \\\n",
       "5      1235797  GCA_000403435.2  Oscillibacter sp. 1-3       1235797   \n",
       "\n",
       "  assembly_level assembly_status     assembly_name refseq_category  \\\n",
       "5       Scaffold         current  Osci_bact_1-3_V1            <NA>   \n",
       "\n",
       "   number_of_contigs  contig_n50  contig_l50  genome_coverage  \\\n",
       "5                 43      271557           6             88.0   \n",
       "\n",
       "   total_sequence_length  total_ungapped_length   gaps  bases_per_contig  \n",
       "5                4467686                4430205  37481     103028.023256  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def select_best_representative(df, consider_coverage=None):\n",
    "    \"\"\"\n",
    "    Select the best representative genome from a genome dataset report.\n",
    "    \"\"\"\n",
    "    if consider_coverage is None:\n",
    "        consider_coverage = True\n",
    "    taxa_dfs = []\n",
    "    # convert columns to correct data types\n",
    "    df = df.astype(\n",
    "        {\n",
    "            \"query_taxid\": \"int\",\n",
    "            \"genome_taxid\": \"int\",\n",
    "            \"number_of_contigs\": \"int\",\n",
    "            \"total_sequence_length\": \"int\",\n",
    "            \"total_ungapped_length\": \"int\",\n",
    "            \"contig_n50\": \"int\",\n",
    "            \"contig_l50\": \"int\",\n",
    "            # \"genome_coverage\": \"float\",\n",
    "        }\n",
    "    )\n",
    "    df[\"gaps\"] = df[\"total_sequence_length\"] - df[\"total_ungapped_length\"]\n",
    "    df[\"bases_per_contig\"] = df[\"total_ungapped_length\"] / df[\"number_of_contigs\"]\n",
    "\n",
    "    if consider_coverage:\n",
    "        # if genome coverage is NA remove it from the dataframe\n",
    "        df = df.dropna(subset=[\"genome_coverage\"]).copy()\n",
    "\n",
    "    # Select the representative genome with the highest genome coverage, least gaps and least contig_l50 and highest contig_n50.\n",
    "    for _, grouped_taxa_df in df.groupby(\"query_taxid\"):\n",
    "        rows = grouped_taxa_df.shape[0]\n",
    "        if rows == 1:\n",
    "            taxa_dfs.append(grouped_taxa_df.iloc[0])\n",
    "            continue\n",
    "\n",
    "        # if a taxon has any 'Complete Genome's then shortlist those genomes to be considered for a representative genome.\n",
    "        if any(grouped_taxa_df[\"assembly_level\"] == \"Complete Genome\"):\n",
    "            grouped_taxa_df = grouped_taxa_df.query(\n",
    "                \"assembly_level == 'Complete Genome'\"\n",
    "            ).copy()\n",
    "\n",
    "        # Select the representative genome with the highest genome coverage, least gaps and least contig_l50 and highest contig_n50.\n",
    "        grouped_taxa_df = grouped_taxa_df.sort_values(\n",
    "            [\"bases_per_contig\", \"contig_n50\", \"contig_l50\", \"genome_coverage\", \"gaps\"],\n",
    "            ascending=[False, False, True, False, True],\n",
    "        )\n",
    "\n",
    "        # append the best representative genome to the taxa_dfs list\n",
    "        taxa_dfs.append(grouped_taxa_df.iloc[0])\n",
    "\n",
    "    return pd.DataFrame(taxa_dfs)\n",
    "\n",
    "\n",
    "select_best_representative(genbank_genomes_df).reset_index(drop=True).sample()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_taxid</th>\n",
       "      <th>genome_accession</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_taxid</th>\n",
       "      <th>assembly_level</th>\n",
       "      <th>assembly_status</th>\n",
       "      <th>assembly_name</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>number_of_contigs</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>genome_coverage</th>\n",
       "      <th>total_sequence_length</th>\n",
       "      <th>total_ungapped_length</th>\n",
       "      <th>gaps</th>\n",
       "      <th>bases_per_contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>446660</td>\n",
       "      <td>GCF_000478885.1</td>\n",
       "      <td>Adlercreutzia equolifaciens DSM 19450</td>\n",
       "      <td>1384484</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM47888v1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1</td>\n",
       "      <td>2862526</td>\n",
       "      <td>1</td>\n",
       "      <td>33.1</td>\n",
       "      <td>2862526</td>\n",
       "      <td>2862526</td>\n",
       "      <td>0</td>\n",
       "      <td>2.862526e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>46506</td>\n",
       "      <td>GCF_000154525.1</td>\n",
       "      <td>Bacteroides stercoris ATCC 43183</td>\n",
       "      <td>449673</td>\n",
       "      <td>Scaffold</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM15452v1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>23</td>\n",
       "      <td>587129</td>\n",
       "      <td>3</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>4009829</td>\n",
       "      <td>4009229</td>\n",
       "      <td>600</td>\n",
       "      <td>1.743143e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>33038</td>\n",
       "      <td>GCF_009831375.1</td>\n",
       "      <td>[Ruminococcus] gnavus ATCC 29149</td>\n",
       "      <td>411470</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM983137v1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1</td>\n",
       "      <td>3549191</td>\n",
       "      <td>1</td>\n",
       "      <td>150.0</td>\n",
       "      <td>3549191</td>\n",
       "      <td>3549191</td>\n",
       "      <td>0</td>\n",
       "      <td>3.549191e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1547</td>\n",
       "      <td>GCF_016728785.1</td>\n",
       "      <td>Thomasclavelia ramosa</td>\n",
       "      <td>1547</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM1672878v1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1</td>\n",
       "      <td>3543719</td>\n",
       "      <td>1</td>\n",
       "      <td>751.0</td>\n",
       "      <td>3543719</td>\n",
       "      <td>3543719</td>\n",
       "      <td>0</td>\n",
       "      <td>3.543719e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>907</td>\n",
       "      <td>GCF_003006415.1</td>\n",
       "      <td>Megasphaera elsdenii</td>\n",
       "      <td>907</td>\n",
       "      <td>Complete Genome</td>\n",
       "      <td>current</td>\n",
       "      <td>ASM300641v1</td>\n",
       "      <td>representative genome</td>\n",
       "      <td>1</td>\n",
       "      <td>2566193</td>\n",
       "      <td>1</td>\n",
       "      <td>137.0</td>\n",
       "      <td>2566193</td>\n",
       "      <td>2566193</td>\n",
       "      <td>0</td>\n",
       "      <td>2.566193e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     query_taxid genome_accession                            genome_name  \\\n",
       "113       446660  GCF_000478885.1  Adlercreutzia equolifaciens DSM 19450   \n",
       "64         46506  GCF_000154525.1       Bacteroides stercoris ATCC 43183   \n",
       "48         33038  GCF_009831375.1       [Ruminococcus] gnavus ATCC 29149   \n",
       "22          1547  GCF_016728785.1                  Thomasclavelia ramosa   \n",
       "7            907  GCF_003006415.1                   Megasphaera elsdenii   \n",
       "\n",
       "     genome_taxid   assembly_level assembly_status assembly_name  \\\n",
       "113       1384484  Complete Genome         current    ASM47888v1   \n",
       "64         449673         Scaffold         current    ASM15452v1   \n",
       "48         411470  Complete Genome         current   ASM983137v1   \n",
       "22           1547  Complete Genome         current  ASM1672878v1   \n",
       "7             907  Complete Genome         current   ASM300641v1   \n",
       "\n",
       "           refseq_category  number_of_contigs  contig_n50  contig_l50  \\\n",
       "113  representative genome                  1     2862526           1   \n",
       "64   representative genome                 23      587129           3   \n",
       "48   representative genome                  1     3549191           1   \n",
       "22   representative genome                  1     3543719           1   \n",
       "7    representative genome                  1     2566193           1   \n",
       "\n",
       "    genome_coverage  total_sequence_length  total_ungapped_length  gaps  \\\n",
       "113            33.1                2862526                2862526     0   \n",
       "64             <NA>                4009829                4009229   600   \n",
       "48            150.0                3549191                3549191     0   \n",
       "22            751.0                3543719                3543719     0   \n",
       "7             137.0                2566193                2566193     0   \n",
       "\n",
       "     bases_per_contig  \n",
       "113      2.862526e+06  \n",
       "64       1.743143e+05  \n",
       "48       3.549191e+06  \n",
       "22       3.543719e+06  \n",
       "7        2.566193e+06  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomes_df = pd.concat(\n",
    "    [\n",
    "        select_best_representative(refseq_genomes_df, consider_coverage=False),\n",
    "        select_best_representative(genbank_genomes_df),\n",
    "    ]\n",
    ").reset_index(drop=True)\n",
    "genomes_df.to_csv(genomes_metadata_file, index=False)\n",
    "genomes_df.sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GCF_016889925.1    1\n",
       "GCF_014131755.1    1\n",
       "GCF_020341675.1    1\n",
       "GCF_025149125.1    1\n",
       "GCF_008571395.1    1\n",
       "                  ..\n",
       "GCF_000169235.1    1\n",
       "GCF_018140635.1    1\n",
       "GCF_003475105.1    1\n",
       "GCF_902375065.1    1\n",
       "GCA_028727085.1    1\n",
       "Name: genome_accession, Length: 160, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomes_df[\"genome_accession\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Check to make sure there are no duplicate genome accessions\n",
    "assert all(\n",
    "    genomes_df[\"genome_accession\"].value_counts() == 1\n",
    "), \"Duplicate genome accessions found!\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the genome accessions to a file\n",
    "genomes_df[\"genome_accession\"].to_csv(genome_accessions_file, index=False, header=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download the genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s \"$api_key\" \"$genome_accessions_file\" \"$genomes_zip_file\"\n",
    "\n",
    "datasets download genome accession \\\n",
    "    --api-key $1 \\\n",
    "    --inputfile $2 \\\n",
    "    --no-progressbar \\\n",
    "    --filename $3\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "trees",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
